src/status_im2/contexts/add_new_contact/events.cljs

```diff
@@ -6,114 +6,133 @@
             [status-im.ethereum.ens :as ens]
             [status-im.ethereum.stateofus :as stateofus]
             [status-im.native-module.core :as status]
-            [status-im2.navigation.events :as navigation]
-            [status-im2.utils.validators :as validators]
+            [status-im2.utils.validators :as v]
             [status-im.utils.utils :as utils]))
 
+;; 1. Domain model
+;; - start with empty 'schema' and progressively add/update to it
+;; - persist this data structure to app-db
+;; 2. Composability https://www.youtube.com/watch?v=kuNxVXnmjHA
+;; - pure functions that transform maps
+;; 3. Tests
+;; - focus on the core decision-making functions first (validate-contact)
+;; - followed by the (hopefully mostly empty) event handlers
+;; - followed by int test (add-contact-test) for status-go & downstream events
+;; 4. Testing event handlers/callbacks
+;; - remove all logic from callback: callback only dispatches
+;; - tests replace the callback with a fn that just returns args
+
+(defn init-contact ; in app-db as [:contacts/new-identity]
+  ([]
+   {:user-public-key nil ; user's public key (not the contact)
+    :input           nil ; raw user input (untrimmed)
+    :id              nil ; public-key|compressed-key|ens
+    :type            nil ; :empty|:public-key|:compressed-key|:ens
+    :ens             nil ; id.eth|id.ens-stateofus
+    :public-key      nil ; public-key (from decompression or ens resolution)
+    :state           nil ; :empty|:invalid|:decompress-key|:resolve-ens|:valid
+    :msg             nil ; keyword i18n msg
+   })
+  ([kv] (-> (init-contact) (merge kv))))
+
+(defn ->id
+  [{:keys [input] :as contact}]
+  (let [regex         #"^https?://join.status.im/u/(.+)"
```

Comment:
should this regex be in a variable? perhaps we could make it's purpose even clearer with a helpful name too

Reply:
Agree

Reply:
Agreed. Will do. Thanks!

Reply:
done